Alternative method to get popular times. Fix bugs in get_circle_centers. Cleanup. Tests.

frinkelpi opened the pull request:
- Fix various issues in get_circle_centers (infinite loop, wrong use of radius). Use geopy instead.
- Popular times: Moved to another file. Random user agent. Use requests instead of urllib.
- Popular times: Other method using Google Maps with the exact place ID. It should also be possible to easily retrieve the current popularity.
- Popular times: Add option to output popular times as datetime: popularity dictionary.
- Test script.
- Updated README
- TODO file

m-wrzr commented:
Hey,
sry for the late reply, and thanks for the PR!
I checked out the code and could you elaborate on some of the points?

- Fix various issues in get_circle_centers (infinite loop, wrong use of radius). Use geopy instead.

Where are the bugs in the current version of get_circle_centers, how do you get the infinite loop and what is the wrong use for the radius? The circle centers for the radar search seem fine, see the current and new version respectively below.
The coordinates are created in`populartimes.get(key, ["bar"], (48.132751, 11.565834), (48.142417,11.584288),radius=180, n_threads=1))`

<img width="967" alt="cover_current" src="https://user-images.githubusercontent.com/6058447/32445136-2a16d150-c305-11e7-8c5e-a5f79edeb820.png">
<img width="933" alt="cover_new" src="https://user-images.githubusercontent.com/6058447/32445138-2a41707c-c305-11e7-9cc0-31c331f67d5c.png">

- Popular times: Moved to another file. Random user agent. Use requests instead of urllib.

We ran into some problems when using random user agents and not turning off the ssl context, did you experience sth. similar in some situations?
If I run tests/test.py via command line I get the following error, but I'm not sure where this is coming from.

`Error occurred during loading data. Trying to use cache server https://fake-useragent.herokuapp.com/browsers/0.1.8
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/urllib/request.py", line 1254, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/http/client.py", line 1107, in request
    self._send_request(method, url, body, headers)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/http/client.py", line 1152, in _send_request
    self.endheaders(body)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/http/client.py", line 1103, in endheaders
    self._send_output(message_body)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/http/client.py", line 877, in send
    self.connect()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/http/client.py", line 1261, in connect
    server_hostname=server_hostname)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/ssl.py", line 385, in wrap_socket
    _context=self)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/ssl.py", line 760, in __init__
    self.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/ssl.py", line 996, in do_handshake
    self._sslobj.do_handshake()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/ssl.py", line 641, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:719)
`

- Popular times: Other method using Google Maps with the exact place ID. It should also be possible to easily retrieve the current popularity.

There seems to be an error while parsing the data.
`Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python3/3.6.1/Frameworks/Python.framework/Versions/3.6/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/wrzr/Desktop/pop-pr/populartimes/crawler.py", line 91, in worker_detail
    get_detail(item)
  File "/Users/wrzr/Desktop/pop-pr/populartimes/crawler.py", line 108, in get_detail
    popularity, rating, rating_n = get_populartimes(searchterm, detail["place_id"])
  File "/Users/wrzr/Desktop/pop-pr/populartimes/populartimes.py", line 13, in get_populartimes
    return PopularTimesMaps(explicitDatetimes).get(place_id)
  File "/Users/wrzr/Desktop/pop-pr/populartimes/populartimes.py", line 118, in get
    pop = re.findall(reg, t)[0]
IndexError: list index out of range
`

frinkelpi commented:
Thank you for the reply.

**About `get_circle_centers`**

If I recall correctly, there was:
1. a permutation between lower/upper and/or east/west and/or north/south in the function with respect to the documentation, which gave an infinite loop in the while. I added a computation of the corners beforehand to avoid that.
2. In `tmp += (0.25 / r) * (radius / math.pi)` and `lower[1] += (0.25 / r) * (radius / math.pi) / math.cos(lower[00] * math.pi / radius`, I suspect that `radius / math.pi` should be `180 / math.pi` and that `0.25` should be a parameter indicating the offset (see e.g. [here](https://stackoverflow.com/a/7478827/194594)). To make sure and be more precise, I replaced the code with a call to [geopy](https://pypi.python.org/pypi/geopy).

**About random user agents and SSL validation**
I suspect a configuration issue with the validation of SSL certificates on your machine, it works well on my end. Otherwise, there is always the possibility to add `verify=False` to `requests.get`.

**About popular times through Maps**

You're right, I had forgotten to handle the absence of "popular time" data. This should be corrected with the latest commit. Also, I didn't intend to switch the default method to the new one, `get_populartimes` has been changed accordingly.

m-wrzr commented:
**About get_circle_centers:**
I checked the code again and you are completely right, that's a bug from my side, after adding a adaptable radar search radius, previously there were always 0.25 km used to cover the search area.

One question towards the number of circles needed to cover the area:
`P = geopy.distance.VincentyDistance(meters=radius / 2).destination(point=P, bearing=0)  # Go North
`
Why do you divide the radius by 2? If the radar search radius and the distance between two points is the same we already cover the area, no need to add additional centers.

**Other:**
I get following error when running the unittests:
` File "/Users/wrzr/Desktop/populartimes-master/populartimes/crawler.py", line 131, in get_detail
    if params["all_places"] or len(detail_json["populartimes"]) > 0:
TypeError: object of type 'NoneType' has no len()`

There were updates in the main brach regarding the possibility to get the current popularity, so merging the PR would be a mess. Can I just add you as a contributer, and we only update the get_circle_center code since yours is correct/better for now, until the rest runs smootly?
